### ISSUE TITLE
Remove dependency on syn

### ISSUE THREAD

**Issue by shepmaster** [shepmaster/fuzzy-pickles#97]
We parse Rust code, so we should figure out how to bootstrap ourselves.

**Comment by shepmaster** [comment]
My current WIP is:

1. Run cargo-expand to see the results of our current syn-based `derive`.
1. Extract all the derivations to new files.
1. `cfg` out all the derivations and replace them with the new files.
1. Add a modified Cargo.toml that points to the same source but provides a different crate name and sets the `cfg` feature.
1. Modify the dependency chain to be fuzzy-pickles->fuzzy-pickles-derive->fuzzy-pickles-bootstrap.

This has the downside of essentially compiling the same source code twice, which eats up twice the time.

**Comment by shepmaster** [comment]
An alternate is to completely abandon user-time code generation and simply check-in the expanded code. This has some downsides as well — namely us as developers will has a worse experience.

**Comment by MSleepyPanda** [comment]
What about implementing `ToTokens` from `quote`?

I don't know how to feel about the circular dependency. Whilst interesting to explore, `syn` is ubiquitous and fast to compile... why not stay with it?

**Comment by shepmaster** [comment]
> What about implementing `ToTokens` from quote?

Yeah, that's something I've been thinking about a little as well, although it'd be distinct from this issue. The "problem" is that everything in fp boils down to an `Extent`, so you have to re-combine the original string and the `Extent` to get a string, which is a bit annoying.

> syn is ubiquitous 

In a magic future world where I steal away all the users of syn, it would be silly to *require* two Rust parsers be compiled.

> and fast to compile

Sometimes — when people compile syn, it's usually a very limited subset, only good for `derive`. If you compile the whole thing, it's noticeably slower. That's another avenue to explore; having a subset of code available.

> why not stay with it

Pride, reducing dependencies, removing redundant dependencies, etc.
